Physics/EvtGen/DalitzKinematics.cpp

```diff
@@ -50,200 +51,75 @@ DalitzKinematics::DalitzKinematics(
-    return true;
+  for (size_t i = 0; i < Events.size(); ++i) {
+    double s2 = (qAB[i] + qBC[i] + qCA[i] - mA[i] * mA[i] - mB[i] * mB[i] -
+                 mC[i] * mC[i]);
 
-  return false;
+    if (s2 < M2)
+      Evts.push_back(Events[i]);
+  }
+  return Evts;
 }
 
-DataPoint DalitzKinematics::convert(const Event &event) const {
-
-  double mA, mB, mC, qAB, qBC, qCA;
-
-  mA = event.ParticleList[0].mass();
-  mB = event.ParticleList[1].mass();
-  mC = event.ParticleList[2].mass();
-  qAB = (event.ParticleList[0].fourMomentum() +
-         event.ParticleList[1].fourMomentum())
-            .invariantMass();
-  qBC = (event.ParticleList[1].fourMomentum() +
-         event.ParticleList[2].fourMomentum())
-            .invariantMass();
-  qCA = (event.ParticleList[2].fourMomentum() +
-         event.ParticleList[0].fourMomentum())
-            .invariantMass();
-
-  /*FourMomentum cms;
-  for (auto s : sys.GetRecoilState())
-    cms += event.particle(s).fourMomentum();
-
-  FourMomentum finalA, finalB;
-  for (auto s : sys.GetFinalStates().at(0))
-    finalA += event.particle(s).fourMomentum();
-
-assert(sys.GetFinalStates().size() == 2 &&
- "DalitzKinematics::convert() | More then two particles.");
-
-
-for (auto s : sys.GetFinalStates().at(1))
-finalB += event.particle(s).fourMomentum();
-
-// Four momentum of the decaying resonance
-FourMomentum resP4 = finalA + finalB;
-double mSq = resP4.invMassSq();
-
-// Calculate sum of final states four momenta
-cms += resP4;
-
-if (mSq <= limits.first) {
-// We allow for a deviation from the limits of 10 times the numerical
-// precision
-if (ComPWA::equal(mSq, limits.first, 10))
-mSq = limits.first;
-else
-throw BeyondPhsp("HelicityKinematics::convert() |"
-               " Point beypond phase space boundaries!");
-}
-if (mSq >= limits.second) {
-// We allow for a deviation from the limits of 10 times the numerical
-// precision
-if (ComPWA::equal(mSq, limits.second, 10))
-mSq = limits.second;
-else
-throw BeyondPhsp("HelicityKinematics::convert() |"
-               " Point beypond phase space boundaries!");
-}
+ComPWA::Data::DataSet
+DalitzKinematics::convert(const std::vector<Event> &Events) const {
+
+  ComPWA::Data::DataSet Dataset;
+
+  std::vector<double> mA, mB, mC, qAB, qBC, qCA, weights;
+  for (auto const &event : Events) {
+    mA.push_back(event.ParticleList[0].mass());
+    mB.push_back(event.ParticleList[1].mass());
+    mC.push_back(event.ParticleList[2].mass());
+    qAB.push_back((event.ParticleList[0].fourMomentum() +
+                   event.ParticleList[1].fourMomentum())
+                      .invariantMass());
+    qBC.push_back((event.ParticleList[1].fourMomentum() +
+                   event.ParticleList[2].fourMomentum())
+                      .invariantMass());
+    qCA.push_back((event.ParticleList[2].fourMomentum() +
+                   event.ParticleList[0].fourMomentum())
+                      .invariantMass());
+    weights.push_back(event.Weight);
+  }
```

Comment:
Can this still handel events which are outside phase space due to numerical instabilities? Or is this not a problem anymore?

Reply:
How do you mean? The logic did not change here. Also before the conversion is always performed. If an Event is outside the phsp boundary, the kinematic variables returned here will show that and `isWithinPhsp()` checked exactly that.

Originally exceptions were thrown inside convert, but this was changed a while ago because parallelstl does not handle throwing code (for good reason).

This was with Event based processing in the Kinematics though. Now such a feature could theoretically be inserted again, because convert does not need to run inside a parallelstl loop anymore (pstl can be used inside convert though, which would be perfectly fine).
But as for the current code: A generator should only create events inside phsp. If there are some troubles you can run the Events through the `reduceToPhsp()` method. And convert will simply convert